presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -336,7 +337,8 @@ public PlanWithProperties visitRowNumber(RowNumberNode node, PreferredProperties
                             .mergeWithParent(preferredProperties));
 
             // TODO: add config option/session property to force parallel plan if child is unpartitioned and window has a PARTITION BY clause
-            if (!child.getProperties().isStreamPartitionedOn(node.getPartitionBy())) {
+            if (!child.getProperties().isStreamPartitionedOn(node.getPartitionBy())
```

Comment:
You should also add support for `TopNRowNumberNode` if it doesn't work

Reply:
Nice catch!